### PR TITLE
Issue #4998: Add columnCharIndex field to LocalizedMessage

### PR DESCRIPTION
**Pull request by timurt**
https://github.com/checkstyle/checkstyle/issues/4998

### CODE REVIEW

**Comment by MEZk** [comment]
@romani @rnveach 
I think that we should suppress TC violation. The fluent interface immutable builder should be implemented (in separate issue) to avoid such amount of constructor arguments, because the class become hard to support.

**Comment by timurt** [comment]
@MEZk 

> I think that we should suppress TC violation. The fluent interface immutable builder should be implemented (in separate issue) to avoid such amount of constructor arguments, because the class become hard to support.

I noticed that following contstructor is used only inside test files and no usage in source files.
```
public LocalizedMessage(int lineNo,
                            int columnNo,
                            String bundle,
                            String key,
                            Object[] args,
                            String moduleId,
                            Class<?> sourceClass,
                            String customMessage)
```

Maybe removing it and rewriting all usages can fix this violation

**Comment by MEZk** [comment]
@timurt 
We cannot just remove it. It is a part of api package. Some projects can rely on it.

**Comment by rnveach** [comment]
I think we should look over these constructors in another issue and see which ones we really need, or if we can rewrite any. The only ones we can't remove are ones that external utils like maven need that haven't updated in a while.

**Comment by romani** [comment]
@timurt ,

from TC
>  LocalizedMessage LocalizedMessage has too many constructors (constructor count = 6) 

this need to resolved by suppression of inspection on this file. Please do.
Suppression reason comment : "we cannot easily remove old c-tors from API classes, they have to stay for some time"

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java [hunk 1]
```diff
@@ -246,7 +246,21 @@ public final void setTabWidth(int tabWidth) {
      * @param args arguments to format
      */
     public final void log(DetailAST ast, String key, Object... args) {
-        log(ast.getLineNo(), ast.getColumnNo(), key, args);
+        final int col = 1 + CommonUtils.lengthExpandedTabs(
+                getLines()[ast.getLineNo() - 1], ast.getColumnNo(), tabWidth);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: please explain this +1 and -1 in comment. It looks scary without comment.

**Comment by romani** [comment]
we released 8.2 version, please rebase all your PRs to our latest master to avoid CI problems

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java (hunk 1 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: done

**Comment by romani** [comment]
@rnveach , please do final review.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java [hunk 2]
```diff
@@ -240,13 +240,40 @@ public final void setTabWidth(int tabWidth) {
     /**
      * Helper method to log a LocalizedMessage.
      *
+     * <pre>
+     *     col = 1 + CommonUtils.lengthExpandedTabs(
+                getLines()[ast.getLineNo() - 1], ast.getColumnNo(), tabWidth)
+     * </pre>
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: @romani Doesn't it seem weird we are duplicating and then documenting code in Javadoc?
javadoc should be user friendly description of method and nuances. If code in method changes, I'm sure no one will remember to update this specific javadoc.

To me it seems this documentation should be comments on the actual code inside the method.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java [hunk 3]
```diff
@@ -43,9 +43,12 @@
  * message.properties files. The underlying implementation uses
  * java.text.MessageFormat.
  *
+ *
+ *
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java]: remove unnecessary empty lines.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java (hunk 2 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: ....... ok, agree, lets remove this code block.

Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java (hunk 3 above)
Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java [hunk 4]
```diff
@@ -240,13 +240,33 @@ public final void setTabWidth(int tabWidth) {
     /**
      * Helper method to log a LocalizedMessage.
      *
+     * {@code CommonUtils.lengthExpandedTabs} returns column number considering tabulation
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: As I stated in previous review, I think this comment should be an in-method comment.
From Javadoc view, there is no `CommonUtils.lengthExpandedTabs` in method definition, so this only makes sense if user reads source code too.

**Comment by romani** [comment]
TC should be resolved:
```
src/main/java/com/puppycrawl/tools/checkstyle/api
 LocalizedMessage.java (1)
51: LocalizedMessage Redundant suppression
```

**Comment by MEZk** [comment]
@timurt 
Ping

**Comment by timurt** [comment]
@MEZk @rnveach 

IDEA inspection throws "Redundant Suppression" message, when 3 suppressions are defined inside "@noinspection" tag.

Created issue on YouTrack
https://youtrack.jetbrains.com/issue/IDEA-178473

As temporal solution "ConstructorWithTooManyParameters" suppression was moved from class to every constructor

**Comment by romani** [comment]
all CIs passed.

@rnveach , please finalize review.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java (hunk 4 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: @timurt Are we going to do this?

Comment (timurt) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: done

**Comment by rnveach** [comment]
Once CI passes we can merge, unless someone has a complaint.

**Comment by MEZk** [comment]
@rnveach @romani 
I'm ok with the changes, fell free to merge.

**Comment by MEZk** [comment]
@rnveach @romani 
[Same error](https://travis-ci.org/checkstyle/checkstyle/jobs/272131294#L2955) is in [master](https://travis-ci.org/checkstyle/checkstyle/jobs/272149359#L2939) branch.

**Comment by romani** [comment]
It is #5078 , just recently fixed.

**Pull request merged**